### DISCUSSION TITLE
Big `.git` folder

### DISCUSSION THREAD

**Discussion by McPatate** [git-lfs/git-lfs#5294] [Q&A] [answered]
Hello, I've noticed that when adding files to a git repository after running `git lfs install` and having the file types tracked in `.gitattributes`, the `.git` folder has a size equal to the sum of the files in the directory.
Looking more closely, the blobs are stored in the `.git/lfs/objects/...` folder and are a copy of the objects.

Is this expected behavior ? When having big repos (think over a TiB), this can be cumbersome and feels like a waste of space. It may be that there's just some sort of symbolic link, but usually when running `ls` you have that information, when here you do not.
I'm basing my question on running `du -h -d 1 .` inside the git folder.

**Comment by chrisd8088** [comment, marked as answer]
Hey, thanks for the question.  Yes, this is expected behaviour.  To keep the size of the `.git/lfs/objects` directory under some limit, you can use the `git lfs prune` [command](https://github.com/git-lfs/git-lfs/blob/91bac118ed88bc858612f4df69c35c342332b427/docs/man/git-lfs-prune.adoc), which has a variety of options.

You might also be to use the `git lfs dedup` [command](https://github.com/git-lfs/git-lfs/blob/91bac118ed88bc858612f4df69c35c342332b427/docs/man/git-lfs-dedup.adoc) if you have an operating system and filesystem which support it.

**Reply by McPatate** [reply]
Ok, thanks for your answer ! Is it lfs that creates the copies ? What are they needed for ?

I'll check `git lfs dedup`.

**Reply by TonyCongqianWang** [reply]
git lfs dedup is exatctly what I was looking for! I also have the problem of using up too much memory. But unfortunatly not all of my systems support deduplication. Luckily the system that doesnt support it only needs the files as data and does not modify them.

So I have an Idea based on the same principle: Give the option to checkout/clone a read-only version of the lfs files. Instead of using copy-on-write, it uses symbolic links and writing should be forbidden - if not possible otherwise this might need elevated rights, I don't know if this is safe, but maybe warn the user and still give the option.

**Reply by mwasplund** [reply]
I really like the idea of a git checkout that pulls all the LFS files in a read only mode and generates links back to a single copy of the file in the lfs object store. This would solve my issues of needing to clone the repo for CI builds where I only want to use the files and where I am hitting size limitations because there repo has two copies of every large file effectively doubling my disk requirements.

**Reply by mwasplund** [reply]
Note: this is effectively the workaround here -> #6206

**Comment by chrisd8088** [comment]
> Is it lfs that creates the copies ? What are they needed for ?

Yes, Git LFS stores the full object data in `.git/lfs/objects` whenever it fetches an object from the remote storage.  However, not all objects will be used in the particular revision that is checked out in the working tree.  Since fetching remote objects is expensive, Git LFS caches them so you can check out different revisions, branches, etc. without the expense of re-fetching objects each time.  But the cost of doing that is that the cache may consume more local storage space than desired.  So that's why `git lfs prune` exists, to let you control what is in the cache.

**Reply by McPatate** [reply]
I see, this makes sense ! Thank you for the explanation :)

**Comment by bbartoldson** [comment]
I am using `git lfs pull` to download 40 TB of data from a remote repo, but I don't have space for the git lfs objects and the data (80 TB total because all the data gets duplicated in`.git/lfs/objects/*`). Is there a way to keep the data AND get the assurance that I have grabbed all the data from the repo (which I think `git lfs` provides) without permanently storing the files in `.git/lfs/objects/*`?

For example, currently I have 20 TB of data downloaded, and I am using 40 TB of disk space because of git lfs objects. I want to download the next 20 TB of data, but I will run out of disk space (I only have 50 TB of space total) in the process. My plan is to remove the git lfs objects for the existing data to free up space for new data, but this introduces two concerns:

1. Once I have all the data and have deleted all their associated git lfs objects, will git be able to assure me that I have all the data downloaded? For example, I would hope that running `git restore --source=HEAD :/` would suggest that everything is downloaded.
2. Will all of the data be usable without the associated git lfs objects? I know I may lose some git functionality by not have these objects, but I am only concerned about git's ability to tell me that I have downloaded everything.

**Reply by LiberiFatali** [reply]
@bbartoldson Did you find a solution?

**Reply by chrisd8088** [reply]
Hey, sorry I didn't notice this question earlier.

One thing to note here is that Git itself can't answer the question of whether your Git LFS objects have been downloaded or not, because it only sees the Git LFS pointer files in the actual Git repository contents.  There are some Git LFS commands and options which might help you out, though.

I'm not entirely sure how your repository is structured, but one thing you might try would be something like:
```
$ GIT_LFS_SKIP_SMUDGE=1 git clone <url> repo
$ cd repo
$ git lfs ls-files
...
$ git lfs pull -I <path1>
$ git lfs pull -I <path2>
```

The `GIT_LFS_SKIP_SMUDGE=1` option tells the Git LFS client not to download any Git LFS objects, so you'll just have the Git repository contents and you'll see Git LFS pointer files in place of your any Git LFS objects.  For any such objects in `HEAD`, the `git lfs ls-files` command should list them, with a `-` instead of a `*` to show that only the pointer file resides in the working tree and Git LFS cache, and there's no copy of the object yet.

You can then selectively download specific paths using the `-I` option to `git lfs pull`.  Once you've pulled down the objects you want, you can delete those and their copies in `.git/lfs/objects` (perhaps by using `git lfs prune`, or just by deleting those files directly), and proceed to download other Git LFS objects.

That all said, if you have Git LFS objects which are not in the `HEAD` commit's tree, but only exist at previous points in your Git commit history, or in other branches, you'll have to adjust your workflow to account for that (e.g., by using `git lfs pull <ref>` to find the Git LFS pointer files in specific commits or references).  Or you could use `GIT_LFS_SKIP_SMUDGE=1 git checkout` to check out specific commits or refs without immediately pulling the Git LFS objects in them, and then pull any specific Git LFS objects you wanted after that.

**Reply by LiberiFatali** [reply]
Thank you @chrisd8088, that helps.

**Reply by bbartoldson** [reply]
Yes, this is helpful, thank you!

**Comment by pmmontanari** [comment]
I think my question falls into here. I am using Git LFS to track simulation files. There are three files that are important to me, and that are very big in general: 

1. The simulation project file, around 200 Mb
2. The results file, around 1 Gb
3. The simulation restart file, around 10 Gb

I added all of these types of files by using the classic `git-lfs track "*.ext"`, with the respective file extensions. As of now, I have 4 branches and about 10 commits, but not all commits have results or simulation restart files. It is an estimate, but let's say that I expect my git repo in the whole to be around 50 Gb.

Today, while running a simulation, I ran out of disk space. I went to see how much space my repo was using: about 230 Gb, of which about 90 Gb were in `.git/lfs/objects` and about 140 Gb were in `.git/fs/temp`.

In my panic to try to sort things out I deleted manually all files in `.git/fs/temp` and searched, one by one, the objects that represented a simulation restart file - basically, the ones that I found with more than 4 Gb in size (I arbitrarily assumed these represented a simulation restart file).

I just now saw about the `git-lfs prune` command, and I will start using it more regularly, as it deleted some obsolete objects as I can see.

Anyway, what I want to understand is: why this problem happened? More precisely,

- When/Why are the files in `.git/fs/temp` created?
- Why were more file versions in `.git/fs/objects` than what I have commited?
- What are "good practices" to avoid this problem? E.g. When/With which frequency should I use `git-lfs prune`?

I don't know if these questions make sense, as I am very noob in terms of Git LFS. I am a bit afraid of continuing using Git LFS after this happened, but I really want to as it seems to be the de-facto solution I was searching to version control numerical simulation data, at least for me.

Regards

Edit: typos

**Reply by chrisd8088** [reply]
There are a few options, first of all, by which you can adjust what `git lfs prune` will consider as candidates to be pruned.  Both the `git-lfs-prune(1)` manual [page](https://github.com/git-lfs/git-lfs/blob/9811573f7b571bc0cafa64ffd5fdd7a6681ba722/docs/man/git-lfs-prune.adoc#recent-files) and the `git-lfs-config(5)` manual [page](https://github.com/git-lfs/git-lfs/blob/9811573f7b571bc0cafa64ffd5fdd7a6681ba722/docs/man/git-lfs-config.adoc#prune-settings) have sections which describe those configuration options.  So you can make the command prune more aggressively, so long as you've pushed objects to a Git remote, since `git lfs prune` will never prune objects if it can't be certain there's a remote copy.

As for the `.git/lfs/tmp` directory, every Git LFS command [runs](https://github.com/git-lfs/git-lfs/blob/9811573f7b571bc0cafa64ffd5fdd7a6681ba722/git-lfs.go#L35) a cleanup step at the end which should delete any temporary files older than [one hour](https://github.com/git-lfs/git-lfs/blob/9811573f7b571bc0cafa64ffd5fdd7a6681ba722/fs/cleanup.go#L75-L79).  If you see files in that directory older than that, you should be able to just run any simple Git LFS command, like `git lfs env`, and they'll be removed.
